jwst/outlier_detection/_fileio.py

```diff
@@ -20,13 +14,53 @@ def save_median(median_model, make_output_path, asn_id=None):
     median_model : ~jwst.datamodels.ImageModel
         The median ImageModel or CubeModel to save
     '''
-    default_suffix = "_outlier_i2d.fits"
-    if asn_id is None:
-        suffix_to_remove = default_suffix
+    median_model = ImageModel(median_data)
+    median_model.update(example_model)
+    median_model.meta.wcs = median_wcs
+    output_path = intermediate_output_path(example_model, "median", make_output_path)
+    _save_to_path(median_model, output_path, suffix="median")
+
+
+def save_drizzled(drizzled_model, make_output_path):
+    expected_tail = "outlier_?2d.fits"
+    suffix = drizzled_model.meta.filename[-len(expected_tail):-5]
+    output_path = intermediate_output_path(drizzled_model, suffix, make_output_path)
```

Comment:
I am trying to understand `intermediate_output_path` and put a breakpoint before this line. For the test file I'm using (running outlier detection with save_intermediate_results=True). I see `drizzled_model.meta.filename` is `jw03516001001_02101_00001_nrca1_a3001_outlier_i2d.fits`. However the output path ends up being `jw03516001001_02101_00001_nrca1_a3001_a3001_outlier_i2d.fits`.
What's the need for intermediate_output_path? It seems to be introducing a bug here.

Reply:
The reason for the additional function is that `make_output_path` doesn't play nicely when the drizzled model is input.  Giving `input_model.meta.filename = jw01309022001_04102_00001_nrs2_outlier_s2d.fits` to `make_output_path(input_model.meta.filename, suffix="outlier_s2d") you get `jw01309022001_04102_00001_nrs2_outlier_outlier_s2d.fits`.  For median, `make_output_path(input_model.meta.filename, suffix="median") gives you `jw01309022001_04102_00002_nrs2_s200b1_outlier_median.fits`.

Not sure why this happens for your test data but not the regression test dataset for which I constructed the test.  That association has a non-empty asn_id, but it ends up in the filename zero times.

Reply:
At your prodding I did figure out a new way to do this making heavier use of `make_output_path`.  It's still a bit hacky, basically just string-replacing `"_outlier_"` with `"_"`.

One blocker was that `make_output_path` is not well documented.  There are no useful examples [here](https://jwst-pipeline.readthedocs.io/en/stable/jwst/stpipe/devel_io_design.html#make-that-filename-step-make-output-path) or in the docstring for customizing filenames, and the docstring indicates that `components` is a "dictionary of string replacements" but it seems like it gets interpreted as a string.

Can you point me to the test you ran, so I can ensure that my change is fixing it?

Reply:
I believe the reason `make_output_path` produces `jw01309022001_04102_00001_nrs2_outlier_outlier_s2d.fits` is that `outlier_s2d` is not an understood suffix but `s2d` is. It then strips `s2d` and adds the provided suffix `outlier_s2d`.

The filenames should contain the `asn_id` (otherwise they can produce conflicts see https://github.com/spacetelescope/jwst/pull/8418).

Reply:
I think any association with an `asn_id` should cause the double `asn_id`. I just checked and the data subset used for profiling described on https://innerspace.stsci.edu/pages/viewpage.action?spaceKey=SCSB&title=ModelLibrary+Performance shows the issue when run with `strun jwst.outlier_detection.OutlierDetectionStep asn.json  --save_intermediate_results=True`.

Yeah `make_output_path` is not well documented and very convoluted but if it can do the job it would allow this PR to not add another layer of complexity on top (with `intermediate_output_path`). If `make_output_path` isn't sufficient I'd suggest we look at if the step needs it at all (I think it's mostly if not entirely for the intermediate products which don't quite follow normal products which `make_output_path` is designed to handle).

Reply:
Yes I see that.  Guess it's too late to change the decision to make suffixes that contain underscores :)

I think the `asn_id` either never was appended properly for certain modes, or that behavior has regressed.  I don't see a test added as part of that PR, and it appears that the only place in the regression test suite where the intermediate results are saved (prior to this PR) is for LRS TSO data.  Right now on main, just setting `save_intermediate_results=True` for the pipeline run in test_nirspec_fs_spec3.py, I get an output filename of `jw01309022001_04102_00001_nrs2_outlier_s2d.fits`.  The input association has `"asn_id": "o022"`, and yeah, the 022 is in there, but the docs are telling me that number is the observation number.

Reply:
@braingram let me know how the file naming looks in the most recent version.  The changes should simplify things: I believe there's no longer a need to pass `asn_id` around the step, and instead `make_output_path` is updated using `functools.partial` just before calls to `save_median` in order to suppress additional appends of the `asn_id` to filenames that already have it.  I tested this on the names coming out of the data subset used for profiling, and the `outlier_i2d`, `median`, and `blot` files all have exactly one instance of the asn id in their filename.